incomplete_cooperative/regret.py

```diff
@@ -0,0 +1,148 @@
+        self.regret = np.zeros((self.number_of_regret_minimizers, self.number_of_coalitions), dtype=RMValue)
+
+    def get_metacoalition_id(self, metacoalition: Iterable[Coalition]) -> int:
+        """Properly get an id of a metacoalition."""
+        coalition_player_ids = self.coalitions_to_player_ids[
+            [x.id for x in metacoalition if len(x) not in [0, 1, 2**self.number_of_players]]]
+        return np.sum(2 ** coalition_player_ids)
+
+    def get_actions_distribution(self, past_actions: Iterable[Coalition] | int) -> np.ndarray:
+        """Get the action predicted by the strategy.
+
+        If past_actions is an int, it is considered a metacoalition id.
+        Otherwise, it is considered as a list of coalitions of the original game.
+        """
+        metacoalition = past_actions if isinstance(past_actions, int) else self.get_metacoalition_id(past_actions)
+        rank = self.meta_id_to_rank[metacoalition]
+        if sum(self.regret[rank]) == 0:
+            return np.ones(self.number_of_coalitions) / self.number_of_coalitions  # TODO: not all viable!
+        return self.regret[rank] * (self.regret[rank] > 0) / sum(self.regret[rank] * self.regret[rank] > 0)
+
+    def apply_regret(self, bottom_losses: np.ndarray, used_actions: list[list[Coalition]]) -> None:
+        """Scan every possible action sequence, and update the regret of all the regret minimizers.
+
+        Arguments:
+            bottom_losses: the loss of the bottom layer of the game tree.
+                    this should be an array of size (number_of_coalitions,).
+            used_actions: the list of list of actions used to generate the bottom_losses.
+        """
+        used_metacoalition_ranks = [self.meta_id_to_rank[self.get_metacoalition_id(x)]
+                                    for x in used_actions]
+        losses = np.zeros((self.number_of_regret_minimizers, self.number_of_coalitions),
+                          dtype=RMValue)
+        experienced_losses = np.zeros(self.viable_metacoalitions, dtype=RMValue)
+        experienced_losses[used_metacoalition_ranks] = bottom_losses
+        for i in range(self.number_of_regret_minimizers - 1, -1, -1):
+            metacoalition = self.meta_rank_to_id[i]
+            coalition_pids = list(Coalition(metacoalition).players)
+            next_coalition_pids = [x for x in range(self.number_of_coalitions) if x not in coalition_pids]
+            next_metacoalitions = [Coalition.from_players(coalition_pids + [x]).id for x in next_coalition_pids]
+            next_metacoalition_ranks = self.meta_id_to_rank[next_metacoalitions]
+            losses[i, next_coalition_pids] = experienced_losses[next_metacoalition_ranks]
+            experienced_losses[i] = (losses[i] * self.get_actions_distribution(int(metacoalition))).sum()
+
+        self.regret += losses[np.arange(self.number_of_regret_minimizers)] - experienced_losses[np.arange(self.number_of_regret_minimizers), np.newaxis]
```

Comment:
self.iteration += 1

Reply:
I added this to the top of the function, as we're starting from zero (is that correct?)